File: Python-Program/Q4.py
import threading

counter = 0

def increment():
    global counter
    for _ in range(100000):
        counter += 1

thread1 = threading.Thread(target=increment)
thread2 = threading.Thread(target=increment)

thread1.start()
thread2.start()

thread1.join
thread2.join()

print("Final counter value:", counter)

File: java-Programs/Q4.java
public class CounterThreadExample {
    static int counter = 0;

    public static void increment() {
        for (int i = 0; i < 100000; i++)
            counter++; // No synchronization causes race conditions
    }

    public static void main(String[] args) {
        Thread t1 = new Thread(() -> increment());
        Thread t2 = new Thread(() -> increment());

        t1.start();
        t2.start();

        try {
            t1.join;
            t2.join; // Syntax error: missing parentheses
        } catch (InterruptedException e) {
            System.out.println("Thread interrupted.");
        }

        System.out.println("Final Counter: " + counter); // Inconsistent results
    }
}